pyomo/solvers/plugins/solvers/gurobi_persistent.py

```diff
@@ -40,55 +41,54 @@ class GurobiPersistent(PersistentSolver, GurobiDirect):
 
     def __init__(self, **kwds):
         kwds['type'] = 'gurobi_persistent'
-        PersistentSolver.__init__(self, **kwds)
-        GurobiDirect._init(self)
+        GurobiDirect.__init__(self, **kwds)
 
         self._pyomo_model = kwds.pop('model', None)
         if self._pyomo_model is not None:
             self.set_instance(self._pyomo_model, **kwds)
 
     def _remove_constraint(self, solver_con):
+        if isinstance(solver_con, self._gurobipy.Constr):
+            if self._solver_model.getAttr('NumConstrs') == 0:
+                self._update()
+            else:
+                name = self._symbol_map.getSymbol(self._solver_con_to_pyomo_con_map[solver_con])
+                if self._solver_model.getConstrByName(name) is None:
+                    self._update()
+        elif isinstance(solver_con, self._gurobipy.QConstr):
+            if self._solver_model.getAttr('NumQConstrs') == 0:
+                self._update()
+            else:
+                try:
+                    qc_row = self._solver_model.getQCRow(solver_con)
+                except self._gurobipy.GurobiError:
+                    self._update()
+        elif isinstance(solver_con, self._gurobipy.SOS):
+            if self._solver_model.getAttr('NumSOS') == 0:
+                self._update()
+            else:
+                try:
+                    sos = self._solver_model.getSOS(solver_con)
+                except self._gurobipy.GurobiError:
+                    self._update()
+        else:
+            raise ValueError('Unrecognized type for gurobi constraint: {0}'.format(type(solver_con)))
         self._solver_model.remove(solver_con)
+        self._needs_updated = True
 
     def _remove_sos_constraint(self, solver_sos_con):
-        self._solver_model.remove(solver_sos_con)
+        self._remove_constraint(solver_sos_con)
+        self._needs_updated = True
 
     def _remove_var(self, solver_var):
+        if self._solver_model.getAttr('NumVars') == 0:
+            self._update()
+        else:
+            name = self._symbol_map.getSymbol(self._solver_var_to_pyomo_var_map[solver_var])
+            if self._solver_model.getVarByName(name) is None:
+                self._update()
         self._solver_model.remove(solver_var)
-
-    def add_var(self, var):
```

Comment:
These methods are implemented in the base class. I originally had special implementations so that `update` could be called. However, that was an inefficient place to be calling `update`. Now, `update` gets called only when it is really needed, which is before removing a variable/constraint that has been added without calling update and before getting a variable/constraint/model attribute.